File: day01/zig/src/main.zig
const std = @import("std");

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const args = try std.process.argsAlloc(allocator);
    const input = try readFileAll(allocator, args[1]);
    try day1(allocator, input);
}

fn readFileAll(allocator: Allocator, path: []const u8) ![]u8 {
    const file = try std.fs.cwd().openFile(path, .{ .mode = .read_only });
    defer file.close();
    return file.readToEndAlloc(allocator, 1024 * 1024);
}

const Allocator = std.mem.Allocator;

pub fn day1(allocator: Allocator, input: []const u8) !void {
    const names = [_][]const u8{
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
    };

    var lines_it = std.mem.splitScalar(u8, input, '\n');
    var total: u32 = 0;
    while (lines_it.next()) |line| {
        const len = line.len;

        var enil: []u8 = try allocator.alloc(u8, len);
        defer allocator.free(enil);
        @memcpy(enil, line);
        std.mem.reverse(u8, enil);

        std.debug.print("Line: {s} <-> {s} \t", .{ line, enil });
        const first_digit = blk: {
            for (0..len) |i| {
                if (std.ascii.isDigit(line[i])) {
                    break :blk (line[i] - '0');
                }
                const slice = line[i..];
                for (names, 0..) |name, name_idx| {
                    if (std.mem.startsWith(u8, slice, name)) {
                        break :blk @as(u32, @intCast(name_idx + 1));
                    }
                }
            }
            break :blk 0;
        };

        const last_digit = blk: {
            for (0..len) |i| {
                if (std.ascii.isDigit(enil[i])) {
                    break :blk (enil[i] - '0');
                }
                const slice = enil[i..];
                for (names, 0..) |name, name_idx| {
                    var name_rev = try allocator.alloc(u8, name.len);
                    defer allocator.free(name_rev);
                    @memcpy(name_rev, name);
                    std.mem.reverse(u8, name_rev);

                    if (std.mem.startsWith(u8, slice, name_rev)) {
                        break :blk @as(u32, @intCast(name_idx + 1));
                    }
                }
            }
            break :blk 0;
        };
        std.debug.print("{d}\t{d}\n", .{ first_digit, last_digit });
        total = total + (first_digit * 10) + last_digit;
    }

    std.debug.print("Day 1: {d}\n", .{total});
}

File: day01/rust/src/main.rs
use std::fs;

fn main() {
    use std::env;

    let input = env::args()
        .nth(1)
        .map(|f| fs::read_to_string(f).unwrap())
        .unwrap();

    let lines = input
        .split("\n")
        .filter(|l| !l.is_empty())
        .collect::<Vec<_>>();

    day1(lines)
}

fn day1(lines: Vec<&str>) {
    let total = lines
        .iter()
        .map(|line| {
            let fd = first_digit(line);
            let ld = last_digit(line);
            println!("GOT <{:?}, {:?}> for {}", fd, ld, line);
            options_to_num(fd, ld)
        })
        .fold(0, |a, i| a + i);
    println!("Total : {}", total);
}

const NAMES: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn options_to_num(a: Option<char>, b: Option<char>) -> u32 {
    match (a, b) {
        (Some(x), Some(y)) => char_to_dec(x) * 10 + char_to_dec(y),
        (Some(x), None) => char_to_dec(x),
        (None, Some(y)) => char_to_dec(y),
        _ => 0,
    }
}

fn char_to_dec(c: char) -> u32 {
    println!("Char '{}'", c);
    (c as u32) - ('0' as u32)
}

fn first_digit(line: &str) -> Option<char> {
    digit(&line.to_string(), &NAMES.map(|n| n.to_string()))
}

fn last_digit(line: &str) -> Option<char> {
    digit(
        &line.chars().rev().collect::<String>(),
        &NAMES.map(|n| n.to_string().chars().rev().collect::<String>()),
    )
}

fn first_name(line: &String, names: &[String; 9]) -> Option<char> {
    names.iter().enumerate().find_map(|(idx, word)| {
        if line.starts_with(word) {
            let num = (idx as u32) + 1 + ('0' as u32);
            char::from_u32(num.try_into().unwrap())
        } else {
            None
        }
    })
}

fn digit(line: &String, names: &[String; 9]) -> Option<char> {
    line.chars()
        .next()
        .filter(|t| t.is_digit(10))
        .or(first_name(line, names))
        .or(if line.is_empty() {
            None
        } else {
            let rest = line.chars().skip(1).collect::<String>();
            digit(&rest, names)
        })
}
